### PR TITLE
Issue 12603 - [CTFE] goto does not correctly call dtors

### PR DESCRIPTION
**Pull request by legrosbuffle**
Fix https://issues.dlang.org/show_bug.cgi?id=12603

In CTFE, goto was not correctly handling struct lifetime: dtors were called wrongfully, or not called at all.

### CODE REVIEW

**Comment by tramker** [comment]
Please put the keyword "pull" to the bugzilla entry and link to this PR and also name this PR properly (Issue NNNN - bug title).

**Comment by legrosbuffle** [comment]
Fixed the naming of the pull request. This fix won't work unfortunately, but at least I have a reduced test case. I'll add a reduced test case for the phobos breakage too.

**Comment by legrosbuffle** [comment]
Can somebody have a look at this one ? @tramker ?

**Comment by 9rnsr** [comment]
OK, I'll look at this tomorrow.

Currently those are the test files for CTFE behaviors:
`test/compilable/interpret3.d`
`test/runnable/interpret.d`
`test/runnable/interpret2.d`

I'd like to collect test cases into few files as far as possible, to make repeated testing easier during CTFE bugfix.

File: test/compilable/b12603.d [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    case 0:
+        auto set = charsets;
+        break;
+    default:
+        break;
+    }
+    return 1;
+}
+
+static assert(structInCaseScope());
```
Comment (9rnsr) [inline test/compilable/b12603.d]: This test code does not look at the result of `S2.dtorCalled`. In other words, what is tested here?

File: test/compilable/b12603.d [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    
+    uint* dtorCalled;
+}
+
+
+auto numDtorCallsByGotoWithinScope()
+{
+    uint dtorCalled;
+    {
+      S s = S(&dtorCalled);
```
Comment (9rnsr) [inline test/compilable/b12603.d]: Please consistently use 4 spaces indent.

File: src/ctfe.h [hunk 3]
```diff
@@ -113,6 +113,7 @@ class CTFEExp : public Expression
     static CTFEExp *gotoexp;
 
     static bool isCantExp(Expression *e) { return e && e->op == TOKcantexp; }
+    static bool isGotoExp(Expression *e) { return e && e->op == TOKgoto; }
```
Comment (9rnsr) [inline src/ctfe.h]: Adding `CTFEExp::isGotoExp` is a kind of refactoring. Please separate the change into a commit.

File: src/interpret.c [hunk 4]
```diff
@@ -1810,22 +1810,46 @@ class Interpreter : public Visitor
...
-            e = interpret(s->body, istate);
-            // Jump into/out from finalbody is disabled in semantic analysis.
-            // and jump inside will be handled by the ScopeStatement == finalbody.
-            result = e;
+            // We were looking for a label, but did not find it, don't interpret the finalbody.
+            result = ex;
             return;
         }
-
-        Expression *ex = interpret(s->body, istate);
```
Comment (9rnsr) [inline src/interpret.c]: These line changes (setting `labelSeeking` etc.) are unrelated with the bugfix. The `if (istate->start)` style is consistent with other interpreter parts. Please don't break consistency.

However, I don't deny the changes for code readability. If you think current code is not well understandable, please open a new PR for refactoring.

**Comment by 9rnsr** [comment]
OK, the core change in `visit(TryFinallyStatement *s)` looks good to me.

File: src/ctfe.h (hunk 3 above)
Comment (legrosbuffle) [inline src/ctfe.h]: Makes sense. I pulle dit out into https://github.com/D-Programming-Language/dmd/pull/4698

Comment (9rnsr) [inline src/ctfe.h]: Thanks.

File: test/compilable/b12603.d (hunk 1 above)
Comment (legrosbuffle) [inline test/compilable/b12603.d]: I investigated a few solutions to fix this bug. On the way to the final one I had one that would give a cantexp on "auto set = charsets;", regardless of the return value.
I like to err more on the side of more unit tests, so I left this one here. I can remove it though, please tell me what you think.

Comment (9rnsr) [inline test/compilable/b12603.d]: I would doubt the necessity of a test-code if it tests nothing. Therefore the old version was not good to me.
However, now you have changed the code as to test the number of destructor call. So I think the new version is good for the bugfix.
Please leave the new version of test code.

**Comment by 9rnsr** [comment]
#4698 was merged. Squash the commits into one, and rebase onto the latest master.

**Comment by legrosbuffle** [comment]
Done, thanks a lot for the review !

**Comment by 9rnsr** [comment]
> Done, thanks a lot for the review !

Oh, sorry. I have a final nitpick: Please change the commit message to "fix Issue 12603 - [CTFE] goto does not correctly call dtors".

When a commit merge is noticed to `issues.dlang.org` website, bugzilla will automatically change the status of issue ticket xxxxx to "RESOLVED FIXED" by seeing the sentence "fix Issue xxxxx" in the commit message.

**Comment by legrosbuffle** [comment]
Done.

**Comment by 9rnsr** [comment]
Thanks!

**Comment by 9rnsr** [comment]
Auto-merge toggled on

**Pull request merged**